File: src/LetterCombination.java
/**
 * 
 * LetterCombination.java
 * Nov 27, 2012
 */
import java.util.*;
/**
 * @author huangsp
 *
 */
public class LetterCombination {
	String[] keyboard = {"0","1","abc","def","ghi","jkl","mno","pqrs","tuv","wxyz"};
	ArrayList<String> list = new ArrayList<String>();
	public ArrayList<String> letterCombinations(String digits) {
		makeCombination(digits, new StringBuffer());
		return list;
    }
	// non-recursive algorithm
	public ArrayList<String> letterCombinations2(String digits) {
        // Start typing your Java solution below
        // DO NOT write main() function
        if (digits == null || digits.length() == 0) return new ArrayList<String>();
        else{
            ArrayList<String> solutionList = new ArrayList<String>();
            int len = digits.length();
            int digit = digits.charAt(0) - '0';
            for (int index = 0; index < keyboard[digit].length(); index ++){
            	String newWord = keyboard[digit].charAt(index) + "";
            	solutionList.add(newWord);
            }
            for (int index = 1; index < len; index ++){
                digit = digits.charAt(index) - '0';
				int size = solutionList.size();
                while(size > 0){
                	size --;
                    String currentWord = solutionList.remove(solutionList.size() - 1);
                    for (int indexD = 0; indexD < keyboard[digit].length(); indexD ++){
                        char newChar = keyboard[digit].charAt(indexD);
                        solutionList.add(0,currentWord + newChar);
                    }
                }
                
            }
            return solutionList;
        }
    }
	private void makeCombination(String digits, StringBuffer out){
		if (digits.length() == 0){
			list.add(out.toString());
		}
		else{
			int num = digits.charAt(0) - '0';
			for (int index = 0; index < keyboard[num].length(); index++){
				out.append(keyboard[num].charAt(index));
				makeCombination(digits.substring(1), out);
				out.setLength(out.length() - 1);
			}
		}
	}
	public static void main(String[] args){
		System.out.println(new LetterCombination().letterCombinations2("23"));
	}
}

File: python/letter_combination.py
'''
Letter Combinations of a Phone Number Total Accepted: 32660 Total Submissions: 123824 My Submissions Question Solution 
Given a digit string, return all possible letter combinations that the number could represent.

A mapping of digit to letters (just like on the telephone buttons) is given below.

Input:Digit string "23"
Output: ["ad", "ae", "af", "bd", "be", "bf", "cd", "ce", "cf"].
'''

class Solution:
    def __init__(self):
        self.keyboard = [ '0',   '1',    'abc', 'def', 'ghi', 
                          'jkl', 'mno', 'pqrs', 'tuv', 'wxyz']
        self.combine_list = list()

    # @return a list of strings, [s1, s2]
    # Recursive is a DFS
    def letterCombinations(self, digits):
        if len(digits) == 0:
            return []
        self.makeLetterCombinations(digits, '')
        return self.combine_list

    # @param digits a string
    # @param word a string
    def makeLetterCombinations(self, digits, word):
        size = len(digits)
        if size == 0:
            self.combine_list.append(word)
            return
        digit = int(digits[0])
        for char in self.keyboard[digit]:
            word += char
            self.makeLetterCombinations(digits[1:], word)
            word = word[:-1]

    # @return a list of strings, [s1, s2]
    # Non-recursive is a BFS
    def letterCombinationsNonRecursive(self, digits):
        size = len(digits)
        if size == 0:
            return []
        digit = int(digits[0])
        combine_list = [ char for char in self.keyboard[digit] ]
        for index in range(1, size):
            digit = int(digits[index])
            chars = self.keyboard[digit]
            previous_list = combine_list
            combine_list = list()
            for char in chars:
                combine_list.extend([ word+char for word in previous_list ])
        return combine_list
    
if __name__ == '__main__':
    solution = Solution()
    print solution.letterCombinationsNonRecursive('23')
